docs/proposals/04-new-core-gardener-cloud-apis.md

```diff
@@ -0,0 +1,718 @@
+    # clusterAutoscaler:
+    #   scaleDownUtilizationThreshold: 0.5
+    #   scaleDownUnneededTime: 30m
+    #   scaleDownDelayAfterAdd: 60m
+    #   scaleDownDelayAfterFailure: 10m
+    #   scaleDownDelayAfterDelete: 10s
+    #   scanInterval: 10s
+    worker:
+    - name: cpu-worker
+      minimum: 3
+      maximum: 5
+    # maxSurge: 1
+    # maxUnavailable: 0
+      machine:
+        type: m5.large
+        image:
+          name: <some-os-name>
+          version: <some-os-version>
+        # providerConfig:
+        #   <some-os-specific-configuration>
+      volume:
+        type: gp2
+        size: 20Gi
+    # providerConfig:
+    #   <some-provider-specific-worker-config>
+    # labels:
+    #   key: value
+    # annotations:
+    #   key: value
+    # taints: # See also https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/
+    # - key: foo
+    #   value: bar
+    #   effect: NoSchedule
+    # kubernetes:
+    #   kubeProxy:
+    #     featureGates:
+    #       SomeKubernetesFeature: true
+    #     mode: IPVS
+    #   kubelet:
+    #     cpuCFSQuota: true
+    #     cpuManagerPolicy: none
+    #     podPidsLimit: 10
+    #     featureGates:
+    #       SomeKubernetesFeature: true
+      zones:
+      - europe-central-1a
+      - europe-central-1b
+  kubernetes:
+    version: 1.15.1
+  # allowPrivilegedContainers: true # 'true' means that all authenticated users can use the "gardener.privileged" PodSecurityPolicy, allowing full unrestricted access to Pod features.
+  dns:
+    # When the shoot shall use a cluster domain no domain and no providers need to be provided - Gardener will
+    # automatically compute a correct domain.
+    domain: crazy-botany.core.my-custom-domain.com
+    providers:
+    - type: aws-route53
+      secretName: my-custom-domain-secret
+      domains:
+        include:
+        - my-custom-domain.com
+        - my-other-custom-domain.com
+        exclude:
+        - yet-another-custom-domain.com
+  extensions:
+  - type: foobar
+  # providerConfig:
+  #   apiVersion: foobar.extensions.gardener.cloud/v1alpha1
+  #   kind: FooBarConfiguration
+  #   foo: bar
+  networking:
+    type: calico
+    pods: 100.96.0.0/11
+    services: 100.64.0.0/13
+    nodes: 10.250.0.0/16
+  # providerConfig:
+  #   apiVersion: calico.extensions.gardener.cloud/v1alpha1
+  #   kind: NetworkConfig
+  #   ipam:
+  #     type: host-local
+  #     cidr: usePodCIDR
+  #   backend: bird
+  # See also: https://github.com/gardener/gardener/blob/master/docs/proposals/03-networking.md
+  maintenance:
+    timeWindow:
+      begin: 220000+0100
+      end: 230000+0100
+    autoUpdate:
+      kubernetesVersion: true
+      machineImageVersion: true
+# hibernation:
+#   enabled: false
+#   schedules:
+#   - start: "0 20 * * *" # Start hibernation every day at 8PM
+#     end: "0 6 * * *"    # Stop hibernation every day at 6AM
+#     location: "America/Los_Angeles" # Specify a location for the cron to run in
+  addons:
```

Comment:
Can we remove this entire field? 
It's a better solution to have a dedicated resource just for those.

For backwards compatibility, we have this field serialized as annotation and the controller is going to create the addon resources separately.

Reply:
As long as we have no sufficient and very well thought-through addon concept I'd not do that. The `addons` section will only contain `kubernetes-dashboard` and `nginx-ingress-controller`, and no more. If you want we can move them to another section `.spec.kubernetes` (?), but let's not create a separate resource for that, at least not now. We don't have any concept and it makes it more difficult for end-users.

Reply:
What about moving it to `spec.extensions`?

Reply:
Do you suggest writing an extension controller for these two? The `.spec.extensions` is only meant for controllers that react on `Extension` resources.
We can think about doing that, but I don't think it makes too much sense. The kubernetes-dashboard is kind of a system component and we might anyway completely remove it in the future (see discussions in #1325). The nginx-ingress-controller is special as well and we probably have to/want to provide one default ingress controller. Moving it out to an extension controller would create the impression that other extensions could/should be written to support other ingress controllers. However, the plan is actually to let end-users install their own controllers based on their requirements, and/or to offer controllers via the [service hub](https://github.com/gardener/hub). So, keeping it inside Gardener itself (and with it a section in the spec) is probably better and easier to maintain.

However, as said, I'm open for further discussions. Either way let's try to not do it with this GEP but separately. We can always deprecate the addons section and migrate to the extensions approach.

Reply:
To me it looks like both fields offer the same functionality. I would like to merge it into one.

Reply:
Are you talking about `.spec.addons` and `.spec.extensions`? Well, not really, they are very different: Every entry in `.spec.extensions` results into an `Extension` CRD object in the shoot namespace in the seed, and it requires an extension controller do reconcile it. The things in `.spec.addons` are well-known addons implemented in in-tree.

As said, we can discuss merge `.spec.addons` into `.spec.extensions`, but this would require implementing extension controllers. Merging `.spec.extensions` into `.spec.addons` is not possible.

Still, I suggest discussing that outside of this GEP. It's additional effort that should not be combined with the already large effort of migrating the APIs. Also, to me it's not yet clear what our plan with the nginx-ingress and kubernetes-dashboard is, especially with regards to the planned hub (cc @vlerenc). While we might remove the dashboard entirely, we probably want to keep nginx-ingress as a default addon. If that's true I doubt that it makes sense putting this into an own extension controller - for the reasons already mentioned above:

> Moving it out to an extension controller would create the impression that other extensions could/should be written to support other ingress controllers. However, the plan is actually to let end-users install their own controllers based on their requirements, and/or to offer controllers via the service hub. So, keeping it inside Gardener itself (and with it a section in the spec) is probably better and easier to maintain.